### PR TITLE
Peer db new serialization

### PR DESCRIPTION
**Pull request by paninaro**
Trello task: https://trello.com/c/bstGWYhf/1381-replace-peer-dbs-with-flat-file

Switches to a new serialization format for the peer data that was previously saved to a sqlite db. The peer data that was previously saved to `peer_db_table.sqlite` and `wallet_peers.sqlite` will now be stored in `peers.dat` and `wallet_peers.dat` respectively.

**Note**: Existing peer data will be migrated to the new format. Existing peer sqlite DBs are left untouched in case users need to downgrade.

AddressManagerStore has been simplified by removing the aiosqlite logic. Deserialization is performed synchronously, while serialization is asynchronous.

`initial-config.yaml` has been updated such that `peer_db_path` and `wallet_peers_path` have been replaced by `peers_file_path` and `wallet_peers_file_path`. While the initial config has been updated, no attempt is made to update existing config.yaml files. By leaving existing config.yaml files as-is, we can support downgrading.

The PeerStoreResolver class has been added to determine the appropriate peer file path given data from the config. The resolved file path is stored in the config object that is passed in during initialization. Currently, this resolved value is not written back to config.yaml, though this is possible if we want to provide a migration option.

Tests have been added to cover PeerStoreResolver and the updates to AddressManagerStore.

### CODE REVIEW

File: chia/server/node_discovery.py [hunk 1]
```diff
@@ -80,6 +82,20 @@ def __init__(
...
+        if self.legacy_peer_db_path is None or not self.legacy_peer_db_path.exists() or self.peers_file_path.exists():
+            return
+        try:
+            self.log.info(f"Migrating legacy peer database from {self.legacy_peer_db_path}")
+            # Attempt to create an AddressManager from the legacy peer database
+            address_manager: Optional[AddressManager] = await AddressManagerSQLiteStore.create(self.legacy_peer_db_path)
+            if address_manager is not None:
+                self.log.info(f"Writing migrated peer data to {self.peers_file_path}")
+                # Write the AddressManager data to the new peers file
+                await AddressManagerStore.serialize(address_manager, self.peers_file_path)
```
Comment (arvidn) [inline chia/server/node_discovery.py:95]: if we get here without an error, shouldn't `self.legacy_peer_db_path` be set to `None`? I suppose this is only called on `start()` so maybe not. However, we don't delete (or move aside) the old DB. How do we record that we've already migrated, to avoid doing it again next time we restart (and clobber our new peer list with an old one)?

File: chia/server/address_manager_sqlite_store.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        if int(metadata.get("new_count", 0)) > 0:
+            return False
+        if int(metadata.get("tried_count", 0)) > 0:
+            return False
+        return True
+
+    @classmethod
+    def _create_address_manager(
+        cls, metadata: Dict[str, str], nodes: List[Node], new_table_entries: List[Table]
+    ) -> AddressManager:
```
Comment (arvidn) [inline chia/server/address_manager_sqlite_store.py:67]: Doesn't the `AddressManager` have a constructor that lets it be constructed from these dicts and lists?
It seems suspicious that you duplicated the code below, rather than moved it. If it's still in its original location, is there a reason you can't use that original copy of it?

**Review by arvidn — COMMENTED** [review]
I think it's important that you add tests for these new features

File: chia/util/write_file.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import os
+import shutil
+
+from pathlib import Path
+from typing import Union
+
+
+def write_file(file_path: Path, data: Union[str, bytes], *, file_mode: int = 0o600):
```
Comment (arvidn) [inline chia/util/write_file.py:8]: aren't there named constants for `0o600`?

I also think there should be unit tests for this function

File: chia/util/write_file.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    """
+    Writes the provided data to a temporary file and then moves it to the final destination.
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(os.path.dirname(file_path), mode=dir_perms, exist_ok=True)
+
+    temp_path: Path = file_path.with_suffix("." + str(os.getpid()))
+    mode: str = "w" if type(data) == str else "wb"
```
Comment (arvidn) [inline chia/util/write_file.py:18]: I would have expected you would need `w+` and `wb+` respectively. what if a previous error left the file (with the pid suffix) around? it would need to be truncated.

There should probably be a test for that.

File: chia/util/write_file.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    with open(os.open(str(temp_path), os.O_CREAT | os.O_EXCL | os.O_WRONLY, file_mode), mode) as f:
```
Comment (arvidn) [inline chia/util/write_file.py:19]: I see, you're using `O_EXCL` here, which would throw an exception if the file already existed, right?
Do we just give up in that case and fail everything?

File: chia/util/write_file.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        f.write(data)
+    try:
+        os.replace(str(temp_path), file_path)
+    except PermissionError:
+        shutil.move(str(temp_path), str(file_path))
```
Comment (arvidn) [inline chia/util/write_file.py:24]: if this call fails, we don't delete the temporary file. should we?

File: chia/util/write_file.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(os.path.dirname(file_path), mode=dir_perms, exist_ok=True)
+
+    temp_path: Path = file_path.with_suffix("." + str(os.getpid()))
+    mode: str = "w" if type(data) == str else "wb"
+    with open(os.open(str(temp_path), os.O_CREAT | os.O_EXCL | os.O_WRONLY, file_mode), mode) as f:
+        f.write(data)
```
Comment (arvidn) [inline chia/util/write_file.py:20]: how do we feel about using `aiofiles` and perform this write asynchronously instead?

i.e.
```
await f.write(data)
```

File: chia/server/address_manager_store.py [hunk 8]
```diff
@@ -10,10 +8,54 @@
...
+    """
+
+    metadata: List[Tuple[str, str]]
+    nodes: List[Tuple[uint64, str]]
+    new_table: List[Tuple[uint64, uint64]]
+
+
+async def makePeerDataSerialization(
+    metadata: List[Tuple[str, Any]], nodes: List[Tuple[int, ExtendedPeerInfo]], new_table: List[Tuple[int, int]]
+) -> PeerDataSerialization:
```
Comment (arvidn) [inline chia/server/address_manager_store.py:37]: is there a reason not to just return `bytes` here directly? as far as I can tell, in all places we call this function we immediately turn the `PeerDataSerialization` object into `bytes`

File: chia/server/address_manager_store.py [hunk 9]
```diff
@@ -145,59 +130,104 @@ async def serialize(self, address_manager: AddressManager):
...
+
         return address_manager
+
+    @classmethod
+    def _read_peers(cls, peers_file_path: Path) -> PeerDataSerialization:
+        """
+        Read the peers file and return the data as a PeerDataSerialization object.
+        """
+        with open(peers_file_path, "rb") as f:
+            return PeerDataSerialization.from_bytes(f.read())
```
Comment (arvidn) [inline chia/server/address_manager_store.py:219]: I would also imagine this would by `async` and use `await f.read()` in an ideal world. Was that considered?

File: chia/util/write_file.py (hunk 6 above)
Comment (arvidn) [reply chia/util/write_file.py:24]: I'm pretty sure that both this and `os.replace()` will fail on windows if some other process (or our process) has the file open. Perhaps it would be reasonable to retry a few times, if you make `write_file()` async, you could async sleep for 0.5 seconds or so and retry a few times, before failing.

**Review by arvidn — COMMENTED** [review]
did you profile this patch? if so, did you measure an improvement?

File: chia/server/node_discovery.py [hunk 10]
```diff
@@ -80,6 +82,20 @@ def __init__(
         if default_port is None and selected_network in NETWORK_ID_DEFAULT_PORTS:
             self.default_port = NETWORK_ID_DEFAULT_PORTS[selected_network]
 
+    async def migrate_address_manager_if_necessary(self) -> None:
+        if self.legacy_peer_db_path is None or not self.legacy_peer_db_path.exists() or self.peers_file_path.exists():
+            return
```
[the remark this answers is not in the document]
Comment (paninaro) [reply chia/server/node_discovery.py:87]: I'd be more comfortable with adding a new instance variable like `self.migrated = True` to allow us to exit early if this method is called more than once. I'll make that change.

File: chia/server/node_discovery.py (hunk 1 above)
Comment (paninaro) [reply chia/server/node_discovery.py:95]: I'll add a `migrated` instance variable to track this within the same run to make things a little tidier. Though, once migrated, the new file exists, so we won't reattempt migration.

File: chia/server/address_manager_sqlite_store.py (hunk 2 above)
Comment (paninaro) [reply chia/server/address_manager_sqlite_store.py:67]: I wanted the migration code to be logically identical to what it was before. The reason for duplicating and not reusing/moving is so that changes can be made in AddressManagerStore going forward without affecting migration.

File: chia/server/address_manager_store.py (hunk 8 above)
Comment (paninaro) [reply chia/server/address_manager_store.py:37]: Sounds reasonable

File: chia/server/node_discovery.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (arvidn) [reply chia/server/node_discovery.py:87]: Makes sense. Is there a mechanism to prevent migrating on every startup though?

File: chia/server/node_discovery.py [hunk 11]
```diff
@@ -80,6 +82,20 @@ def __init__(
...
+        try:
+            self.log.info(f"Migrating legacy peer database from {self.legacy_peer_db_path}")
+            # Attempt to create an AddressManager from the legacy peer database
+            address_manager: Optional[AddressManager] = await AddressManagerSQLiteStore.create(self.legacy_peer_db_path)
+            if address_manager is not None:
+                self.log.info(f"Writing migrated peer data to {self.peers_file_path}")
+                # Write the AddressManager data to the new peers file
+                await AddressManagerStore.serialize(address_manager, self.peers_file_path)
+        except Exception:
+            self.log.exception("Error migrating legacy peer database")
```
[the remark this answers is not in the document]
Comment (arvidn) [reply chia/server/node_discovery.py:97]: I see, I didn't know that

File: chia/server/node_discovery.py (hunk 1 above)
Comment (arvidn) [reply chia/server/node_discovery.py:95]: I see, we should probably check for its existence first then (before the db)

**Comment by arvidn** [comment]
Also, I think we should be careful to not aim *too* high in this patch. It sounds like we have a decent speed-up, and we should probably and this before trying to squeeze more performance out of this. It might not be the lowest hanging fruit anymore.

File: chia/server/address_manager_sqlite_store.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+    @classmethod
+    async def _get_new_table(cls, connection: aiosqlite.Connection) -> List[Table]:
+        cursor = await connection.execute("SELECT node_id, bucket from peer_new_table")
+        entries = await cursor.fetchall()
+        await cursor.close()
+        return [(node_id, bucket) for node_id, bucket in entries]
+
+    @classmethod
+    async def _is_empty(cls, metadata: Dict[str, str]) -> bool:
```
[the remark this answers is not in the document]
Comment (paninaro) [reply chia/server/address_manager_sqlite_store.py:55]: I'll change it. I do generally prefer to group related functions into a class (even if it's stateless), especially if I want to provide some semblance of a public/private interface, but if it's not the python way, that's fine.

[the remark this answers is not in the document]
Comment (arvidn) [reply chia/server/address_manager_sqlite_store.py:55]: I should have added that I'm not an authority on the "python way". I would say it would be a code smell in C++, but maybe it isn't in python.

File: chia/util/write_file.py (hunk 3 above)
Comment (paninaro) [reply chia/util/write_file.py]: Tests added.

Re: The named constants for the permission mask. I strongly prefer the octal representation, which I think is fairly universally understood by devs. The alternative is to explode out a bitmask of similarly named constants like `stat.S_IRUSR | stat.S_IWUSR |  stat.S_IRGRP | stat.S_IWGRP` etc. I find it far more readable to use the octal representation, and less prone to accidentally choose the wrong permission bit(s). The octal representation is interoperable with the named constants.

File: chia/util/write_file.py (hunk 4 above)
Comment (paninaro) [reply chia/util/write_file.py]: I've switched it to use w+ and w+b. Existing file issue has been worked-around by using the tempfile functionality to guarantee a uniquely named temp file.

File: chia/util/write_file.py (hunk 5 above)
Comment (paninaro) [reply chia/util/write_file.py]: Changed to use the tempfile module to guarantee a unique temp file for writing.

File: chia/util/write_file.py (hunk 6 above)
Comment (paninaro) [reply chia/util/write_file.py]: There were some edge cases that caused issues on Windows, as you hinted at. I've added retry logic, and updated the code such that the temp file is removed if the move fails.

Several tests to handle os.replace failures. The tests pass for me on macOS and Windows.

File: chia/util/write_file.py (hunk 7 above)
Comment (paninaro) [reply chia/util/write_file.py]: I had considered aiofiles, but initially wanted to keep the change as small/tidy as possible. That's changed now, so I've pulled in aiofiles.

File: chia/server/address_manager_store.py (hunk 9 above)
Comment (paninaro) [reply chia/server/address_manager_store.py]: Updated to use aiofiles.

**Comment by paninaro** [comment]
> Also, I think we should be careful to not aim _too_ high in this patch. It sounds like we have a decent speed-up, and we should probably and this before trying to squeeze more performance out of this. It might not be the lowest hanging fruit anymore.
> 
> (this is a counter-weight against my comments about making the disk I/O async)

This was my original goal -- to address *just* the serialization format while squeezing some additional performance out at the same time. The data model could likely still use some tweaking to get the footprint down further. As it stands, the streamable-based serialization format is faster at serializing and nominally slower at deserializing (only occurring once during launch) when compared to sqlite. Serialization times were roughly 8x faster when I last checked.

**Comment by arvidn** [comment]
the `aiofiles` dependency caused this:
```
chia/server/address_manager_store.py:1: error: Library stubs not installed for "aiofiles" (or incompatible with Python 3.8)
chia/server/address_manager_store.py:1: note: Hint: "python3 -m pip install types-aiofiles"
chia/server/address_manager_store.py:1: note: (or run "mypy --install-types" to install all missing stub packages)
```

**Comment by paninaro** [comment]
> the `aiofiles` dependency caused this:
> 
> ```
> chia/server/address_manager_store.py:1: error: Library stubs not installed for "aiofiles" (or incompatible with Python 3.8)
> chia/server/address_manager_store.py:1: note: Hint: "python3 -m pip install types-aiofiles"
> chia/server/address_manager_store.py:1: note: (or run "mypy --install-types" to install all missing stub packages)
> ```

Thanks for the pointer. Looks like a couple of the workflow files have their own sections for additional dependencies. I updated setup.py (dev_dependencies), upload-pypi-source.yml (which doesn't sound related, but...) and pre-commit-config.yaml.

**Comment by arvidn** [comment]
it looks like the LGTM issues are legitimate ans should be addressed

**Comment by paninaro** [comment]
> it looks like the LGTM issues are legitimate ans should be addressed

I find it strange that LGTM is complaining about these files now -- I haven't touched them in this PR. I'd rather not touch these files, given that they aren't related to the changes. Post-merge, I'm fine fixing the unused import, though the self-assignment issues identified in `timelord_state.py` should be looked at by someone knowledgable in that area.

**Comment by arvidn** [comment]
I see

**Review by arvidn — DISMISSED** [review]

File: chia/util/files.py [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(file_path.parent, mode=dir_perms, exist_ok=True)
+
+    mode: str = "w+" if type(data) == str else "w+b"
+    temp_file_path: Path
+    async with tempfile.NamedTemporaryFile(dir=file_path.parent, mode=mode, delete=False) as f:
+        temp_file_path = f.name
+        await f.write(data)
```
Comment (arvidn) [inline chia/util/files.py:74]: I think it would be a good idea to add:
```
await f.flush()
```

here, to make sure the data actually hits the disk before swapping the files out, in case of catastrophic failures at the wrong time

File: chia/util/files.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+
+async def write_file_async(file_path: Path, data: Union[str, bytes], *, file_mode: int = 0o600):
+    """
+    Writes the provided data to a temporary file and then moves it to the final destination.
+    """
+
+    dir_perms: int = 0o700
+    # Create the parent directory if necessary
+    os.makedirs(file_path.parent, mode=dir_perms, exist_ok=True)
```
Comment (arvidn) [inline chia/util/files.py:68]: is there a good reason to specify the permission bits here?
Do we specifically want to disallow the user from controlling this a more permissive `umask`?

Comment (paninaro) [reply chia/util/files.py:68]: Having a more secure default seems like the right behavior for the case that the intermediates don't already exist. For use cases that want to ensure something more open, they can create the intermediates before calling this. We could have the directory permissions passed into the function.

File: chia/util/files.py (hunk 13 above)
Comment (paninaro) [reply chia/util/files.py:74]: I don't know that it would make a difference, but maybe. Since this is used in a context manager via `async with` the temp file will be closed before the files are swapped out and any buffers should be flushed in the process. Unless the close is a lie.

File: chia/util/files.py (hunk 14 above)
Comment (arvidn) [reply chia/util/files.py:68]: closing a file does not synchronize with the disk.

Comment (arvidn) [reply chia/util/files.py:68]: ehm. what I mean is `close(fd)` doesn't block until the data has hit the disk. it's not trivial to find authoritative documentation on closing files, but flushing would have to be an `async` function, and there's no `await` call corresponding to closing of the context manager, is there?

File: chia/util/files.py (hunk 13 above)
Comment (arvidn) [reply chia/util/files.py:74]: the [documentation](https://docs.python.org/3/library/io.html#io.IOBase.close) says:

> Flush and close this stream

And the documentation for `flush()` says:

> Flush the write buffers of the stream if applicable.

So, it's not explicit about whether it's also blocks until that flush completes, but it seems reasonable to assume that.

Comment (paninaro) [reply chia/util/files.py:74]: I'll add a call to flush to make it explicit.

**Review by wjblanke — APPROVED** [review]
aok
